### ISSUE TITLE
Promise tests

### ISSUE THREAD

**Issue by ollef** [danr/ts-quickcheck#11]
Hej!

Splendid work on the library! I'm having a great time so far. :) It addresses some issues I had with Jsverify, and additionally has a nice and simple interface.

I ran into a hurdle: many of the tests I want to port are promise-based.

Mocha allows you to test promises with `it`, e.g.:

```typescript
it("promises to do something", async () => ...)
```

Jsverify allows something similar with its `property`.

Would it be possible to add support for promises to `createProperty`, or is it already possible to do this?

Cheers!

**Comment by ollef** [comment]
I'm not too familiar with Tape, but I see that there's a `tape-promise` library. Could that work?

**Comment by danr** [comment]
Hej! Happy to hear that you find it helpful!

There is no support for this right now. I did run into the same problem and I would like to support this too.

Internally all counterexample searches are made with this function

```typescript
export function search<A>(
  g: Gen<A>,
  prop: (a: A, p: Property) => boolean,
  options?: Partial<Options>
): SearchResult<A> 
```

I don't foresee any problems with making an async/promise-based version too:

```typescript
export function asyncSearch<A>(
  g: Gen<A>,
  prop: (a: A, p: Property) => Promise<boolean>,
  options?: Partial<Options>
): Promise<SearchResult<A>>
```

I don't know when I will have time to work on this though.

I think tape can test asynchronous code out of the box, it waits for `t.end()` to be called.

**Comment by ollef** [comment]
Any ideas for doing it without duplicating the whole of `search`?

**Comment by danr** [comment]
I think the `async` notation will force you into a Promise world which you cannot get out of. One could try to make a promise-like thing which is actually the identity monad but even if it works it might now work in future TS versions.

Another way would be to use generators and yield every monadic request. Then two different functions will interpret these requests, one in the promise monad and pure interpretation. The TS typechecking for generators that return is quite non-existent afaik. I'll give this a try now nevertheless.

**Comment by danr** [comment]
Nope, generators won't help (directly) since it's not possible to `yield` under a lambda:

```typescript
async function test(k: (h: () => void) => void) {
  k(() => yield 1)
}
```

(Here I want the call to `k` to yield to `test`, not to itself)

**Comment by danr** [comment]
I could isolate it to evaluating the tree in the two different monads. It avoids some code duplication but not all: https://github.com/danr/proptest/commit/a9e20735891a82e2b7b87600dc1091e6d2ecac24

The tests pass for sync tests, I haven't tested async at all yet.

**Comment by ollef** [comment]
Nice, thanks for looking into it! I'll try it out and report back to you when I get the chance to do so.
